Tests/ApolloCacheDependentTests/WatchQueryTests.swift

```diff
@@ -3,701 +3,988 @@ import XCTest
+    }
+  }
+  
+  func testWatchedQueryRefetchesFromServerAfterOtherQueryUpdatesListWithIncompleteObject() throws {
+    client.store.cacheKeyForObject = { $0["id"] }
+    
+    let watchedQuery = HeroAndFriendsNamesWithIDsQuery()
+    
+    let resultObserver = makeResultObserver(for: watchedQuery)
+    
+    let watcher = GraphQLQueryWatcher(client: client, query: watchedQuery, resultHandler: resultObserver.handler)
+    addTeardownBlock { watcher.cancel() }
+    
+    runActivity("Initial fetch from server") { _ in
+      let serverRequestExpectation = server.expect(HeroAndFriendsNamesWithIDsQuery.self) { request in
+        [
+          "data": [
+            "hero": [
+              "id": "2001",
+              "name": "R2-D2",
+              "__typename": "Droid",
+              "friends": [
+                ["__typename": "Human", "id": "1000", "name": "Luke Skywalker"],
+                ["__typename": "Human", "id": "1002", "name": "Han Solo"],
+                ["__typename": "Human", "id": "1003", "name": "Leia Organa"],
+              ]
+            ]
+          ]
+        ]
+      }
+      
+      let initialWatcherResultExpectation = resultObserver.expectation(description: "Watcher received initial result from server") { result in
+        let graphQLResult = try result.get()
+        XCTAssertEqual(graphQLResult.source, .server)
+        XCTAssertNil(graphQLResult.errors)
         
-        defer {
-          fetching.fulfill()
-        }
+        let data = try XCTUnwrap(graphQLResult.data)
+        XCTAssertEqual(data.hero?.name, "R2-D2")
+        let friendsNames = data.hero?.friends?.compactMap { $0?.name }
+        XCTAssertEqual(friendsNames, ["Luke Skywalker", "Han Solo", "Leia Organa"])
+      }
+      
+      watcher.fetch(cachePolicy: .fetchIgnoringCacheData)
+      
+      wait(for: [serverRequestExpectation, initialWatcherResultExpectation], timeout: defaultWaitTimeout)
+    }
+    
+    runActivity("Fetch other query with list of updated keys from server") { _ in
+      let serverRequestExpectation = server.expect(HeroAndFriendsIDsQuery.self) { request in
+        [
+          "data": [
+            "hero": [
+              "id": "2001",
+              "name": "Artoo",
+              "__typename": "Droid",
+              "friends": [
+                ["__typename": "Human", "id": "1003"],
+                ["__typename": "Human", "id": "1004"],
+                ["__typename": "Human", "id": "1000"],
+              ]
+            ]
+          ]
+        ]
+      }
+      
+      let refetchServerRequestExpectation = server.expect(HeroAndFriendsNamesWithIDsQuery.self) { request in
+        [
+          "data": [
+            "hero": [
+              "id": "2001",
+              "name": "Artoo",
+              "__typename": "Droid",
+              "friends": [
+                ["__typename": "Human", "id": "1003", "name": "Leia Organa"],
+                ["__typename": "Human", "id": "1004", "name": "Wilhuff Tarkin"],
```

Comment:
I had no idea this was Grand Moff Tarkin's real name.

Reply:
I'm more of a Star Trek fan myself, so this reference was completely lost on me :)

Reply:
🖖